hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -1742,6 +1742,69 @@ case class TableFilterIntervals(child: TableIR, intervals: IndexedSeq[Interval],
   }
 }
 
+case class TableGroupWithinPartitions(child: TableIR, n: Int) extends TableIR {
+  lazy val children: IndexedSeq[BaseIR] = Array(child)
+
+  lazy val rowCountUpperBound: Option[Long] = child.rowCountUpperBound
+
+  override lazy val typ: TableType = TableType(
+    child.typ.keyType ++ TStruct(("groupedValues", TArray(child.typ.rowType))),
+    child.typ.key,
+    child.typ.globalType)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): TableIR = {
+    val IndexedSeq(newChild: TableIR) = newChildren
+    TableGroupWithinPartitions(newChild, n)
+  }
+
+  override def execute(ctx: ExecuteContext): TableValue = {
+    val prev = child.execute(ctx)
+    val prevRVD = prev.rvd
+    val prevKeyType = prev.rvd.typ.kType
+
+    val groupedElementsPType = PArray(prevRVD.typ.rowType, false)
+    val rowType = this.typ.rowType.physicalType
```

Comment:
this isn't allowed -- `.physicalType` is about to go away, since it converts from virtual type to physical type. I'll have to think for a moment about what method we should use instead (probably directly construct a new PStruct from the previous key field ptypes and the new array)